src/Bicep.Core/Semantics/DeclarationVisitor.cs

```diff
@@ -461,5 +469,85 @@ public ScopeInfo(LocalScope scope)
 
             public IList<ScopeInfo> Children { get; } = new List<ScopeInfo>();
         }
+
+        private ResultWithDiagnostic<ResourceTypesProviderDescriptor> TryGetProviderDescriptor(ProviderDeclarationSyntax syntax)
+        {
+            // LegacyProviderSpecification will be deprecated in the future
+            if (syntax.Specification is LegacyProviderSpecification legacySpecification)
+            {
+                if (!features.DynamicTypeLoadingEnabled && !features.ProviderRegistryEnabled)
+                {
+                    return new(ResourceTypesProviderDescriptor.CreateBuiltInProviderDescriptor(
+                        legacySpecification.NamespaceIdentifier,
+                        legacySpecification.Version,
+                        syntax.Alias?.IdentifierName));
+                }
+
+                return new(x => x.ExpectedProviderSpecification());
+            }
+
+            if (syntax.Specification is ConfigurationManagedProviderSpecification configSpec && 
+                configuration.ProvidersConfig.IsSysOrBuiltIn(configSpec.NamespaceIdentifier))
```

Comment:
Should we also condition this to return a provider descriptor only when `features.DynamicTypeLoadingEnabled` ?

Reply:
Handling of "built-in" providers feels like it fits more under the "extensibility" feature, so I left this alone.